src/kern/npf_alg_pptp.c

```diff
@@ -0,0 +1,763 @@
+
+	/* note: only the outbound NAT is supported */
+	if (di != PFIL_OUT || proto != IPPROTO_TCP || npc->npc_l4.tcp == NULL ||
+			  npc->npc_l4.tcp->th_dport != htons(PPTP_SERVER_PORT))
+		return false;
+
+	/* Associate ALG with translation entry. */
+	npf_nat_setalg(nt, pptp_alg.alg_pptp_tcp, 0);
+	return true;
+}
+
+/*
+ *
+ */
+static int
+npfa_pptp_gre_establish_gre_conn(npf_cache_t *npc, int di,
+		  struct pptp_gre_con *gre_con, npf_nat_t *pptp_tcp_nt)
+{
+	npf_conn_t *con = NULL;
+	int ret;
+
+	NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "establishing a new pptp gre connection: client call_id %hu, "
+			  "server call_id %hu, orig client call_id %hu\n",
+			  gre_con->ctx.client_call_id, gre_con->ctx.server_call_id,
+			  gre_con->orig_client_call_id);
+
+	/* establish new gre connection state */
+	con = npf_conn_establish(npc, di, true);
+	if (con == NULL) {
+		NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+				  "failed to establish pptp gre connection\n");
+		return ENOMEM;
+	}
+	else {
+		NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+				  "new pptp gre connection is established, flags %u\n",
+				  con->c_flags);
+	}
+
+	NPF_HEX_DUMPCL(NPF_DC_PPTP_ALG, 50,
+			  "gre forw key",
+			  npf_conn_getforwkey(con),
+			  NPF_CONNKEY_V4WORDS * sizeof(uint32_t));
+	NPF_HEX_DUMPCL(NPF_DC_PPTP_ALG, 50,
+			  "gre back key",
+			  npf_conn_getbackkey(con, NPF_CONNKEY_ALEN(npf_conn_getforwkey(con))),
+			  NPF_CONNKEY_V4WORDS * sizeof(uint32_t));
+
+	/*
+	 * Create a new nat entry for created GRE connection.
+	 * Use the same nat policy as the parent PPTP TCP control connection uses.
+	 * Associate created nat entry with the gre connection.
+	 */
+	ret = npf_nat_share_policy(npc, con, pptp_tcp_nt);
+	if (ret)
+		return ret;
+
+	/* associate GRE ALG with the gre connection */
+	npf_nat_setalg(con->c_nat, pptp_alg.alg_pptp_gre, (uintptr_t)gre_con->u64);
+
+	NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "new pptp gre connection's nat %p\n", con->c_nat);
+
+	/* make gre connection active and pass */
+	npf_conn_set_active_pass(con);
+
+	gre_con->flags |= PPTP_ALG_FL_GRE_STATE_ESTABLISHED;
+	return 0;
+}
+
+static uint16_t
+npfa_translated_call_id_get(uint32_t ip)
+{
+	in_port_t port;
+	npf_addr_t addr;
+
+	addr.word32[0] = ip;
+	port = npf_portmap_get_pm(pptp_alg.pm, &pptp_alg.pm_params, sizeof(ip),
+			  &addr);
+
+	NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "pptp alg: get call_id %hu from "
+			  "the pormap ip %u\n", port, ip);
+
+	return (uint16_t)port;
+}
+
+static void
+npfa_translated_call_id_put(uint32_t ip, uint16_t call_id)
+{
+	npf_addr_t addr;
+	addr.word32[0] = ip;
+
+	npf_portmap_put_pm(pptp_alg.pm, sizeof(ip), &addr, (in_port_t)call_id);
+
+	NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "pptp alg: put call_id %hu to "
+			  "the pormap ip %u\n", call_id, ip);
+}
+
+/* lookup for and explicitly expire
+ * the associatiated gre connection state
+ */
+static void
+npfa_pptp_expire_gre_con(npf_t *npf, struct pptp_gre_con *gre_con,
+		  uint32_t client_ip, uint32_t server_ip)
```

Comment:
Even if it's IPv4-only, you I think you should use npf_addr_t (and check or  pass the address length where needed). This also keeps the interfaces uniform.